File: 22.Automated-Testing-Unit-Integration-and-E2E/code/typescript/table_driven_parametrized_tests_5.ts
describe('applyDiscount table driven', () => {
    const cases = [
        { name: "ten percent off 200", subtotal: 200, percent: 10, want: 180 },
        { name: "zero discount", subtotal: 100, percent: 0, want: 100 },
        { name: "full discount", subtotal: 100, percent: 100, want: 0 },
        { name: "rounds down", subtotal: 99, percent: 10, want: 89.1 }
    ];

    it.each(cases)('$name', ({ subtotal, percent, want }) => {
        expect(subtotal - (subtotal * percent / 100)).toBe(want);
    });
});

File: 22.Automated-Testing-Unit-Integration-and-E2E/code/rust/table_driven_parametrized_tests_5.rs
fn apply_discount(cart_total: u32, percent: u32) -> u32 {
    cart_total - (cart_total * percent / 100)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_apply_discount_table() {
        let cases = vec![
            ("ten percent off 200", 200, 10, 180),
            ("zero discount", 100, 0, 100),
            ("full discount", 100, 100, 0),
            ("rounds down", 99, 10, 89),
        ];

        for (name, subtotal, percent, want) in cases {
            assert_eq!(apply_discount(subtotal, percent), want, "{}", name);
        }
    }
}
